proposed/2023/InsecureConnectionsDisableCertificateValidation.md

```diff
@@ -0,0 +1,124 @@
+
+<!-- One-paragraph description of the proposal. -->
+NuGet does not provide much functionality for controlling the decisions of trust against http and https sources. In recent surveys, we have even seen a general sentiment towards using 2 or more sources for their projects. Many of these sources have different security requirements that should be up to the developer to decide on. 
+
+There are two functionalities that need to be added to NuGet tooling to help companies move over to a "HTTPS Everywhere" ideal world.
+
+1. I have a HTTP server and don't want the HTTPS warning/error.
+2. I have a HTTPS server and I want to ignore SSL/TLS certification validation.
+
+In both of these cases, developers would like more fine-grain control to opt-out of these security best practices such in cases where they have their own local http source or are self-signing with a certificate.
+
+## Motivation 
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+When we blogged about an initial plan about ["HTTPS Everywhere" for NuGet](https://devblogs.microsoft.com/nuget/https-everywhere/), we were met with widespread feedback on scenarios we aren't always aware of. The primary scenario is that a company may have a HTTP source on a local network that they would like to remain using based on their other company security practices. There's also still a challenge of obtaining a SSL/TLS certificate which may take longer than the timeline we initially proposed (~3 years).
+
+With this feedback in mind, we would like to ensure that we have a generally agreed upon plan that provides these mechanism for flexibility based on where you are personally on your journey to "HTTPS Everywhere" in the [.NET 8 timeframe given it is a LTS](https://dotnet.microsoft.com/en-us/platform/support/policy). While these mechanisms are not recommended to use long-term and generally are seen as not an ideal security practice, we believe that it is not our place to push specific security ideology and rather make our tools flexible to meet everyone where they are.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+There will be multiple ways to opt-out of these two functionalities described above.
+
+#### Global NuGet.config
+
+A developer may choose to want a global setting applied to all sources in which they may allow insecure connections on any defined HTTP source and disable certificate validation on any defined HTTPS source.
+
+```
+<!-- Disables certification validation for all HTTPS sources and allows insecure connections from all  HTTP sources -->
+<config>
+    <add key="allowInsecureConnections" value="true" />
+    <add key="disableCertificateValidation" value="true" />
```

Comment:
This allows ALL HTTPS problems to be ignored, right? Some other tools allow additional trusted roots so that issues like unexpected CAs, expired certificates, mismatching domain name still cause things to fail. This can be done by providing an alternative or additional cert bundle, as mentioned here https://github.com/NuGet/Home/issues/4387#issuecomment-275274426.

There are several comments around workarounds which leverage modifying the trusted CAs (e.g. `update-ca-certificates`). This is a less drastic solution than just disabling all HTTPS checks.

Reply:
I'd be careful of disabling all validation with a single switch. I'd still want to know about name mismatches and expired certs, but that may end up super complicated. 

Might an interactive experience help? I add a package at the command line, there's an TLS error of Untrusted root, or expired cert or subject mismatch, do you want to Ignore (One Time), Ignore (always), stop

Even better would be ignored just for the presented cert. So if, in the case of expiry, you accept the current cert, it then gets updated, and expires again, when its a new cert, error again. 

So maybe a "trust this cert with this sha256"?

Reply:
My main case is being forced to use https internally because http is apparently being dropped even for servers that aren't even on the internet. Interactivity will not help in our ci/cd pipelines. I just want to disabled any and all checks since the security of the internal server is completely pointless.

Reply:
@robindegen do you have docs to help us understand what you are referring to? Especially on what it is validating? Thank you!

Reply:
Little background for reference: With Conan the remote servers are managed system-wide (not exactly but for the sake of simplicity). They also say you must trust the repository completely when you use conan on it since it is using python and a conan package can run any kind of python code on the system when pulling in a package. There is no sandbox.

When adding a remote server you simply specify verify_ssl=False. What exact things this ignores and allows I don't know but it is enough for my simple use case of simply not having a trusted certificate. I think this is the most common use case when disabling ssl, an internal server that just can be trusted. https://docs.conan.io/en/1.53/reference/commands/misc/remote.html

For nuget I would expect something similar in nuget.conf.

Reply:
Thank you for your feedback. I have redacted the other experiences to scope a quick solution here.